### PR TITLE
feat: Nx.Random.choice

### PR DESCRIPTION
**Pull request by polvalente**

### CODE REVIEW

File: nx/lib/nx/constants.ex [hunk 1]
```diff
@@ -292,6 +292,19 @@ defmodule Nx.Constants do
...
 
+  @doc """
+  Returns the smallest normal floating point number.
+
+  The smallest normal floating point number is the smallest positive value with
+  1 as the leading bit for the mantissa.
+
+  Accepts the same options as `Nx.tensor/2`
+  """
+  def smallest_normal(type, opts \\ []) do
```
Comment (polvalente) [inline nx/lib/nx/constants.ex:303]: this is the name Jax uses, but I'm not sure if we should use `min_normal` to keep with `min_finite`

File: nx/lib/nx/random.ex [hunk 2]
```diff
@@ -601,6 +601,245 @@ defmodule Nx.Random do
...
+      g = -g - Nx.log(p)
+      idx = g |> Nx.argsort() |> Nx.slice_along_axis(0, n_draws, axis: 0)
+
+      result = Nx.take(tensor, idx, axis: axis)
+      {result, k}
+    end
+  end
+
+  @doc """
+  Sample Gumbel random values with given shape and float dtype.
```
Comment (polvalente) [inline nx/lib/nx/random.ex:815]: Should I expose this function, or should we keep it hidden until someone actually needs it?

If it should be exposed, I need to add doctests

File: nx/lib/nx/random.ex [hunk 3]
```diff
@@ -601,6 +601,245 @@ defmodule Nx.Random do
...
+
+    {u, k} = uniform(key, Nx.Constants.smallest_normal(type), 1, shape: shape, type: type)
+    result = -Nx.log(-Nx.log(u))
+
+    {result, k}
+  end
+
+  #   _check_shape("gumbel", shape)
+  #   return -jnp.log(-jnp.log(
+  #       uniform(key, shape, dtype, minval=jnp.finfo(dtype).tiny, maxval=1.)))
```
Comment (polvalente) [inline nx/lib/nx/random.ex:841]:
```suggestion
```

File: nx/lib/nx/type.ex [hunk 4]
```diff
@@ -590,6 +590,30 @@ defmodule Nx.Type do
...
+  def smallest_normal({:bf, 16}) do
+    f32 = smallest_normal({:f, 32})
+
+    if System.endianness() == :little do
+      <<_::16, x::16-bitstring>> = f32
+      x
+    else
+      <<x::16-bitstring, _::16>> = f32
+      x
+    end
```
Comment (polvalente) [inline nx/lib/nx/type.ex:608]: there must be a better way to do this :p

File: nx/lib/nx/constants.ex (hunk 1 above)
Comment (josevalim) [reply nx/lib/nx/constants.ex:303]: I think we had this in the past as `min_positive` or something like that and we removed it because it was no longer used.

Comment (polvalente) [reply nx/lib/nx/constants.ex:303]: https://github.com/elixir-nx/nx/commit/fc9a2e7148a1d54e35fdd6edb38197c924d3e1d0

found it!

Shall I restore it instead?

File: nx/lib/nx/random.ex [hunk 5]
```diff
@@ -601,6 +601,242 @@ defmodule Nx.Random do
...
+      result = Nx.take(tensor, idx, axis: axis)
+      {result, key}
+    else
+      {shuffled, key} = shuffle(key, tensor, axis: axis)
+      result = Nx.slice_along_axis(shuffled, 0, n_draws, axis: axis)
+      {result, key}
+    end
+  end
+
+  defnp choice_probabilities(key, tensor, p, opts \\ []) do
```
Comment (josevalim) [inline nx/lib/nx/random.ex:781]: It seems options are always required, so I think you should do this:

```suggestion
  defn choice(key, tensor, p, opts) do
```

File: nx/lib/nx/random.ex [hunk 6]
```diff
@@ -601,6 +601,242 @@ defmodule Nx.Random do
...
+      >
+      iex> {result, _key} = Nx.Random.choice(k, t, p, samples: 6, replace: false) # without replacement
+      iex> result
+      #Nx.Tensor<
+        s64[6]
+        [3, 1, 2, 5, 4, 0]
+      >
+  """
+  deftransform choice(key, tensor, p), do: choice_probabilities(key, tensor, p)
+  deftransform choice(key, tensor, p, opts), do: choice_probabilities(key, tensor, p, opts)
```
Comment (josevalim) [inline nx/lib/nx/random.ex:727]:
```suggestion
  deftransform choice(key, tensor, p, opts) when is_list(opts), do: choice_probabilities(key, tensor, p, opts)
```

Options seem required...

File: nx/lib/nx/random.ex [hunk 7]
```diff
@@ -601,6 +601,242 @@ defmodule Nx.Random do
...
+        [5, 0, 0, 4, 0, 3]
+      >
+      iex> {result, _key} = Nx.Random.choice(k, t, samples: 6, replace: false) # without replacement
+      iex> result
+      #Nx.Tensor<
+        s64[6]
+        [2, 0, 4, 5, 1, 3]
+      >
+  """
+  deftransform choice(key, tensor), do: choice_uniform(key, tensor, [])
```
Comment (josevalim) [inline nx/lib/nx/random.ex:664]:
```suggestion
```

File: nx/lib/nx/random.ex (hunk 2 above)
Comment (josevalim) [reply nx/lib/nx/random.ex:816]: Sounds fine by me. Should we also add `_split` versions to all of these functions?

File: nx/lib/nx/type.ex (hunk 4 above)
Comment (josevalim) [reply nx/lib/nx/type.ex:608]: https://github.com/elixir-nx/nx/commit/fc9a2e7148a1d54e35fdd6edb38197c924d3e1d0

File: nx/lib/nx/constants.ex (hunk 1 above)
Comment (josevalim) [reply nx/lib/nx/constants.ex:303]: Whatever you prefer!

File: nx/lib/nx/random.ex (hunk 2 above)
Comment (polvalente) [reply nx/lib/nx/random.ex:816]: What does `_split` do?

File: nx/lib/nx/random.ex (hunk 5 above)
Comment (polvalente) [reply nx/lib/nx/random.ex:781]: indeed they are :)
probabilities are not, so I'll still keep `choice_probabilities/choice_uniform` as separate `defnp`s

Comment (polvalente) [reply nx/lib/nx/random.ex]: No need for this actually :)

File: nx/lib/nx/random.ex (hunk 2 above)
Comment (polvalente) [reply nx/lib/nx/random.ex:810]: @josevalim Actually,  this depends on `shuffle`,  which `split`s the key underneath a few times. Does it still make sense to have `choice_split`?

Comment (josevalim) [reply nx/lib/nx/random.ex:810]: Ah, good point. Try to have a `_split` if we can, but if you cannot, no problem!

Comment (polvalente) [reply nx/lib/nx/random.ex:810]: I don't see a point of having `choice_split` because of `shuffle`, but we can have `gumbel_split`

File: nx/lib/nx/constants.ex [hunk 8]
```diff
@@ -292,6 +292,46 @@ defmodule Nx.Constants do
     Nx.tensor(Complex.new(0, 1), opts)
   end
 
+  @doc """
+  Returns a scalar tensor with the maximum finite value for the given type.
+
+  ## Options
```
Comment (polvalente) [inline nx/lib/nx/constants.ex:298]:
```suggestion
  ## Options
  
```

**Review by josevalim — APPROVED** [review]
